File: 2023/go/1_2/main.go
package main

import (
	"bufio"
	"bytes"
	"fmt"
	"os"
	"strconv"
	"strings"
	"unicode"
)

func main() {
	data, err := os.ReadFile("input")
	if err != nil {
		panic(err)
	}

	b := bytes.NewReader(data)
	s := bufio.NewScanner(b)
	s.Split(bufio.ScanLines)

	var sum int64

	for s.Scan() {
		var digit string

		for i := 0; i < len(s.Text()); i++ {
			d, ok := getDigitAsString(s.Text()[i:])
			if ok {
				digit += d
				break
			}
		}

		for i := len(s.Text()) - 1; i >= 0; i-- {
			d, ok := getDigitAsString(s.Text()[i:])
			if ok {
				digit += d
				break
			}
		}

		n, err := strconv.ParseInt(digit, 10, 64)
		if err != nil {
			panic(err)
		}

		sum += n
	}

	fmt.Println(sum)
}

func getDigitAsString(s string) (string, bool) {
	switch {
	case unicode.IsDigit(rune(s[0])):
		return string(s[0]), true
	case strings.Index(s, "one") == 0:
		return "1", true
	case strings.Index(s, "two") == 0:
		return "2", true
	case strings.Index(s, "three") == 0:
		return "3", true
	case strings.Index(s, "four") == 0:
		return "4", true
	case strings.Index(s, "five") == 0:
		return "5", true
	case strings.Index(s, "six") == 0:
		return "6", true
	case strings.Index(s, "seven") == 0:
		return "7", true
	case strings.Index(s, "eight") == 0:
		return "8", true
	case strings.Index(s, "nine") == 0:
		return "9", true
	}

	return "", false
}

File: 2023/rust/1_2/src/main.rs
use std::fs;

fn main() {
    let content = fs::read_to_string("input").expect("file not found");

    let mut sum: u32 = 0;

    for line in content.lines() {
        let mut digit = "".to_string();

        for i in 0..line.len() {
            if let Some(x) = get_digit(&line[i..].to_string()) {
                digit.push(x);
                break;
            }
        }

        for i in (0..line.len()).rev() {
            if let Some(x) = get_digit(&line[i..].to_string()) {
                digit.push(x);
                break;
            }
        }

        if let Ok(digit) = digit.parse::<u32>() {
            sum += digit
        } else {
            panic!("NaN")
        }
    }

    println!("{}", sum)
}

fn get_digit(s: &String) -> Option<char> {
    if s.chars().nth(0).unwrap().is_digit(10) {
        Some(s.chars().nth(0).unwrap())
    } else if s.starts_with("one") {
        Some('1')
    } else if s.starts_with("two") {
        Some('2')
    } else if s.starts_with("three") {
        Some('3')
    } else if s.starts_with("four") {
        Some('4')
    } else if s.starts_with("five") {
        Some('5')
    } else if s.starts_with("six") {
        Some('6')
    } else if s.starts_with("seven") {
        Some('7')
    } else if s.starts_with("eight") {
        Some('8')
    } else if s.starts_with("nine") {
        Some('9')
    } else {
        None
    }
}
